[TEST-FAILURE] ClientExecutionPoolSizeLowTest.testOwnerNodeTerminateWithAsyncOperations & testOwnerNodeTerminateWithAsyncOperations

```
java.lang.AssertionError: the size of the map is not correct expected:<1000> but was:<999>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at com.hazelcast.test.HazelcastTestSupport$7.run(HazelcastTestSupport.java:623)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:737)
	at com.hazelcast.test.HazelcastTestSupport.assertSizeEventually(HazelcastTestSupport.java:620)
	at com.hazelcast.test.HazelcastTestSupport.assertSizeEventually(HazelcastTestSupport.java:616)
	at com.hazelcast.client.io.ClientExecutionPoolSizeLowTest.testOwnerNodeTerminateWithAsyncOperations(ClientExecutionPoolSizeLowTest.java:93)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.maintenance-nightly/com.hazelcast$hazelcast-client/608/testReport/junit/com.hazelcast.client.io/ClientExecutionPoolSizeLowTest/testOwnerNodeTerminateWithAsyncOperations/

--- a/hazelcast-client-new/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java
+++ b/hazelcast-client-new/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java
@@ -39,7 +39,7 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
 
     private static final ILogger LOGGER = Logger.getLogger(ClientExecutionService.class);
     private static final long TERMINATE_TIMEOUT_SECONDS = 30;
-    private final ExecutorService executor;
+    private final ExecutorService userExecutor;
     private final ExecutorService internalExecutor;
     private final ScheduledExecutorService scheduledExecutor;
 
@@ -59,9 +59,9 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
                         throw new RejectedExecutionException(message);
                     }
                 });
-        executor = new ThreadPoolExecutor(executorPoolSize, executorPoolSize, 0L, TimeUnit.MILLISECONDS,
+        userExecutor = new ThreadPoolExecutor(executorPoolSize, executorPoolSize, 0L, TimeUnit.MILLISECONDS,
                 new LinkedBlockingQueue<Runnable>(),
-                new PoolExecutorThreadFactory(threadGroup, name + ".cached-", classLoader),
+                new PoolExecutorThreadFactory(threadGroup, name + ".user-", classLoader),
                 new RejectedExecutionHandler() {
                     public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
                         String message = "Internal executor rejected task: " + r + ", because client is shutting down...";
@@ -76,7 +76,7 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
     }
 
     public <T> ICompletableFuture<T> submitInternal(Runnable runnable) {
-        CompletableFutureTask futureTask = new CompletableFutureTask(runnable, null, getAsyncExecutor());
+        CompletableFutureTask futureTask = new CompletableFutureTask(runnable, null, internalExecutor);
         internalExecutor.submit(futureTask);
         return futureTask;
     }
@@ -87,20 +87,20 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
 
     @Override
     public void execute(Runnable command) {
-        executor.execute(command);
+        userExecutor.execute(command);
     }
 
     @Override
     public ICompletableFuture<?> submit(Runnable task) {
         CompletableFutureTask futureTask = new CompletableFutureTask(task, null, getAsyncExecutor());
-        internalExecutor.submit(futureTask);
+        userExecutor.submit(futureTask);
         return futureTask;
     }
 
     @Override
     public <T> ICompletableFuture<T> submit(Callable<T> task) {
         CompletableFutureTask<T> futureTask = new CompletableFutureTask<T>(task, getAsyncExecutor());
-        internalExecutor.submit(futureTask);
+        userExecutor.submit(futureTask);
         return futureTask;
     }
 
@@ -117,7 +117,7 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
     public ScheduledFuture<?> scheduleAtFixedRate(final Runnable command, long initialDelay, long period, TimeUnit unit) {
         return scheduledExecutor.scheduleAtFixedRate(new Runnable() {
             public void run() {
-                execute(command);
+                executeInternal(command);
             }
         }, initialDelay, period, unit);
     }
@@ -133,29 +133,25 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
 
     @Override
     public ExecutorService getAsyncExecutor() {
-        return executor;
+        return userExecutor;
     }
 
     public void shutdown() {
-        internalExecutor.shutdown();
-        try {
-            boolean success = internalExecutor.awaitTermination(TERMINATE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
-            if (!success) {
-                LOGGER.warning("InternalExecutor awaitTermination could not completed in "
-                        + TERMINATE_TIMEOUT_SECONDS + " seconds");
-            }
-        } catch (InterruptedException e) {
-            LOGGER.warning("Internal Executor await termination is interrupted", e);
-        }
-        scheduledExecutor.shutdownNow();
+        shutdownExecutor("internal", internalExecutor);
+        shutdownExecutor("scheduled", scheduledExecutor);
+        shutdownExecutor("user", userExecutor);
+    }
+
+    private void shutdownExecutor(String name, ExecutorService executor) {
         executor.shutdown();
         try {
             boolean success = executor.awaitTermination(TERMINATE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
             if (!success) {
-                LOGGER.warning("Executor awaitTermination could not completed in " + TERMINATE_TIMEOUT_SECONDS + " seconds");
+                LOGGER.warning(name + " executor awaitTermination could not completed in "
+                        + TERMINATE_TIMEOUT_SECONDS + " seconds");
             }
         } catch (InterruptedException e) {
-            LOGGER.warning("Executor await termination is interrupted", e);
+            LOGGER.warning(name + " executor await termination is interrupted", e);
         }
     }
 }

--- a/hazelcast-client-new/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java
+++ b/hazelcast-client-new/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java
@@ -24,14 +24,15 @@ import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
 import com.hazelcast.client.impl.protocol.ClientMessage;
 import com.hazelcast.client.spi.ClientExecutionService;
 import com.hazelcast.client.spi.ClientInvocationService;
+import com.hazelcast.core.ExecutionCallback;
 import com.hazelcast.core.HazelcastInstanceNotActiveException;
+import com.hazelcast.core.ICompletableFuture;
 import com.hazelcast.core.LifecycleService;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.Logger;
 import com.hazelcast.nio.Address;
 import com.hazelcast.nio.Connection;
 import com.hazelcast.spi.exception.RetryableHazelcastException;
-import com.hazelcast.util.EmptyStatement;
 
 import java.io.IOException;
 import java.util.concurrent.RejectedExecutionException;
@@ -203,10 +204,8 @@ public class ClientInvocation implements Runnable {
             return false;
         }
         beforeRetry();
-
         try {
-            sleep();
-            ((ClientExecutionServiceImpl) executionService).executeInternal(this);
+            rescheduleInvocation();
         } catch (RejectedExecutionException e) {
             if (LOGGER.isFinestEnabled()) {
                 LOGGER.finest("Retry could not be scheduled ", e);
@@ -216,20 +215,34 @@ public class ClientInvocation implements Runnable {
         return true;
     }
 
-    protected void beforeRetry() {
+    private void rescheduleInvocation() {
+        executionService.schedule(new Runnable() {
+            @Override
+            public void run() {
+                ICompletableFuture<?> future = ((ClientExecutionServiceImpl) executionService)
+                        .submitInternal(ClientInvocation.this);
+                future.andThen(new ExecutionCallback() {
+                    @Override
+                    public void onResponse(Object response) {
+                    }
 
+                    @Override
+                    public void onFailure(Throwable t) {
+                        if (LOGGER.isFinestEnabled()) {
+                            LOGGER.finest("Failure during retry ", t);
+                        }
+                        clientInvocationFuture.setResponse(t);
+                    }
+                });
+            }
+        }, RETRY_WAIT_TIME_IN_SECONDS, TimeUnit.SECONDS);
+    }
+
+    protected void beforeRetry() {
     }
 
     protected boolean shouldRetry() {
         return reSendCount.incrementAndGet() < retryCountLimit;
-    }
-
-    private void sleep() {
-        try {
-            Thread.sleep(TimeUnit.SECONDS.toMillis(RETRY_WAIT_TIME_IN_SECONDS));
-        } catch (InterruptedException ignored) {
-            EmptyStatement.ignore(ignored);
-        }
     }
 
     private boolean isBindToSingleConnection() {

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientExecutionServiceImpl.java
@@ -39,7 +39,7 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
 
     private static final ILogger LOGGER = Logger.getLogger(ClientExecutionService.class);
     private static final long TERMINATE_TIMEOUT_SECONDS = 30;
-    private final ExecutorService executor;
+    private final ExecutorService userExecutor;
     private final ExecutorService internalExecutor;
     private final ScheduledExecutorService scheduledExecutor;
 
@@ -59,9 +59,9 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
                         throw new RejectedExecutionException(message);
                     }
                 });
-        executor = new ThreadPoolExecutor(executorPoolSize, executorPoolSize, 0L, TimeUnit.MILLISECONDS,
+        userExecutor = new ThreadPoolExecutor(executorPoolSize, executorPoolSize, 0L, TimeUnit.MILLISECONDS,
                 new LinkedBlockingQueue<Runnable>(),
-                new PoolExecutorThreadFactory(threadGroup, name + ".cached-", classLoader),
+                new PoolExecutorThreadFactory(threadGroup, name + ".user-", classLoader),
                 new RejectedExecutionHandler() {
                     public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
                         String message = "Internal executor rejected task: " + r + ", because client is shutting down...";
@@ -76,7 +76,7 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
     }
 
     public <T> ICompletableFuture<T> submitInternal(Runnable runnable) {
-        CompletableFutureTask futureTask = new CompletableFutureTask(runnable, null, getAsyncExecutor());
+        CompletableFutureTask futureTask = new CompletableFutureTask(runnable, null, internalExecutor);
         internalExecutor.submit(futureTask);
         return futureTask;
     }
@@ -87,20 +87,20 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
 
     @Override
     public void execute(Runnable command) {
-        executor.execute(command);
+        userExecutor.execute(command);
     }
 
     @Override
     public ICompletableFuture<?> submit(Runnable task) {
         CompletableFutureTask futureTask = new CompletableFutureTask(task, null, getAsyncExecutor());
-        internalExecutor.submit(futureTask);
+        userExecutor.submit(futureTask);
         return futureTask;
     }
 
     @Override
     public <T> ICompletableFuture<T> submit(Callable<T> task) {
         CompletableFutureTask<T> futureTask = new CompletableFutureTask<T>(task, getAsyncExecutor());
-        internalExecutor.submit(futureTask);
+        userExecutor.submit(futureTask);
         return futureTask;
     }
 
@@ -117,7 +117,7 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
     public ScheduledFuture<?> scheduleAtFixedRate(final Runnable command, long initialDelay, long period, TimeUnit unit) {
         return scheduledExecutor.scheduleAtFixedRate(new Runnable() {
             public void run() {
-                execute(command);
+                executeInternal(command);
             }
         }, initialDelay, period, unit);
     }
@@ -133,29 +133,25 @@ public final class ClientExecutionServiceImpl implements ClientExecutionService 
 
     @Override
     public ExecutorService getAsyncExecutor() {
-        return executor;
+        return userExecutor;
     }
 
     public void shutdown() {
-        internalExecutor.shutdown();
-        try {
-            boolean success = internalExecutor.awaitTermination(TERMINATE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
-            if (!success) {
-                LOGGER.warning("InternalExecutor awaitTermination could not completed in "
-                        + TERMINATE_TIMEOUT_SECONDS + " seconds");
-            }
-        } catch (InterruptedException e) {
-            LOGGER.warning("Internal Executor await termination is interrupted", e);
-        }
-        scheduledExecutor.shutdownNow();
+        shutdownExecutor("internal", internalExecutor);
+        shutdownExecutor("scheduled", scheduledExecutor);
+        shutdownExecutor("user", userExecutor);
+    }
+
+    private void shutdownExecutor(String name, ExecutorService executor) {
         executor.shutdown();
         try {
             boolean success = executor.awaitTermination(TERMINATE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
             if (!success) {
-                LOGGER.warning("Executor awaitTermination could not completed in " + TERMINATE_TIMEOUT_SECONDS + " seconds");
+                LOGGER.warning(name + " executor awaitTermination could not completed in "
+                        + TERMINATE_TIMEOUT_SECONDS + " seconds");
             }
         } catch (InterruptedException e) {
-            LOGGER.warning("Executor await termination is interrupted", e);
+            LOGGER.warning(name + " executor await termination is interrupted", e);
         }
     }
 }

--- a/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java
+++ b/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java
@@ -26,14 +26,15 @@ import com.hazelcast.client.impl.client.RetryableRequest;
 import com.hazelcast.client.spi.ClientExecutionService;
 import com.hazelcast.client.spi.ClientInvocationService;
 import com.hazelcast.client.spi.EventHandler;
+import com.hazelcast.core.ExecutionCallback;
 import com.hazelcast.core.HazelcastInstanceNotActiveException;
+import com.hazelcast.core.ICompletableFuture;
 import com.hazelcast.core.LifecycleService;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.Logger;
 import com.hazelcast.nio.Address;
 import com.hazelcast.nio.Connection;
 import com.hazelcast.spi.exception.RetryableHazelcastException;
-import com.hazelcast.util.EmptyStatement;
 
 import java.io.IOException;
 import java.util.concurrent.RejectedExecutionException;
@@ -228,8 +229,7 @@ public class ClientInvocation implements Runnable {
         }
 
         try {
-            sleep();
-            ((ClientExecutionServiceImpl) executionService).executeInternal(this);
+            rescheduleInvocation();
         } catch (RejectedExecutionException e) {
             if (LOGGER.isFinestEnabled()) {
                 LOGGER.finest("Retry could not be scheduled ", e);
@@ -239,12 +239,27 @@ public class ClientInvocation implements Runnable {
         return true;
     }
 
-    private void sleep() {
-        try {
-            Thread.sleep(TimeUnit.SECONDS.toMillis(RETRY_WAIT_TIME_IN_SECONDS));
-        } catch (InterruptedException ignored) {
-            EmptyStatement.ignore(ignored);
-        }
+    private void rescheduleInvocation() {
+        executionService.schedule(new Runnable() {
+            @Override
+            public void run() {
+                ICompletableFuture<?> future = ((ClientExecutionServiceImpl) executionService)
+                        .submitInternal(ClientInvocation.this);
+                future.andThen(new ExecutionCallback() {
+                    @Override
+                    public void onResponse(Object response) {
+                    }
+
+                    @Override
+                    public void onFailure(Throwable t) {
+                        if (LOGGER.isFinestEnabled()) {
+                            LOGGER.finest("Failure during retry ", t);
+                        }
+                        clientInvocationFuture.setResponse(t);
+                    }
+                });
+            }
+        }, RETRY_WAIT_TIME_IN_SECONDS, TimeUnit.SECONDS);
     }
 
     private boolean isBindToSingleConnection() {

